src/main/java/com/artipie/asto/FileStorage.java

```diff
@@ -0,0 +1,143 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.asto;
+
+import com.jcabi.log.Logger;
+import hu.akarnokd.rxjava3.jdk8interop.CompletableInterop;
+import hu.akarnokd.rxjava3.jdk8interop.SingleInterop;
+import io.reactivex.rxjava3.core.Completable;
+import io.reactivex.rxjava3.core.Flowable;
+import io.reactivex.rxjava3.core.Single;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Flow;
+import java.util.stream.Collectors;
+import org.reactivestreams.FlowAdapters;
+
+/**
+ * Simple storage, in files.
+ *
+ * @since 0.1
+ */
+public final class FileStorage implements Storage {
+
+    /**
+     * Where we keep the data.
+     */
+    private final Path dir;
+
+    /**
+     * Ctor.
+     *
+     * @param path The path to the dir
+     */
+    public FileStorage(final Path path) {
+        this.dir = path;
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(final String key) {
+        return (CompletableFuture<Boolean>) Single.fromCallable(
+            () -> Files.exists(Paths.get(this.dir.toString(), key))
+        ).to(SingleInterop.get());
+    }
+
+    @Override
+    public CompletableFuture<Collection<String>> list(final String prefix) {
+        return (CompletableFuture<Collection<String>>) Single.fromCallable(
+            () -> {
+                final String separator = FileSystems.getDefault().getSeparator();
+                if (!prefix.endsWith(separator)) {
+                    throw new IllegalArgumentException(
+                        String.format(
+                            "The prefix must end with '%s': \"%s\"",
+                            prefix,
+                            separator
+                        )
+                    );
+                }
+                final Path path = Paths.get(this.dir.toString(), prefix);
+                final int dirnamelen = path.toString().length() - prefix.length() + 1;
+                final Collection<String> keys = Files.walk(path)
+                    .filter(Files::isRegularFile)
+                    .map(Path::toString)
+                    .map(p -> p.substring(dirnamelen))
+                    .collect(Collectors.toList());
+                Logger.info(
+                    this,
+                    "Found %d objects by the prefix \"%s\" in %s by %s: %s",
+                    keys.size(), prefix, this.dir, path, keys
+                );
+                return keys;
+            }).to(SingleInterop.get());
+    }
+
+    @Override
+    public CompletableFuture<Void> save(final String key, final Flow.Publisher<Byte> content) {
+        final Completable result = Flowable.fromPublisher(FlowAdapters.toPublisher(content))
```

Comment:
@Sammers21 the result can be inlined like in all other methods: `return Flowable.fromPublisher(...)`